Fix the issue that len(indices) and num_samples might not be equal

sunjq1 opened the pull request:
### What changes were proposed in this pull request?

 Modified the definition of `total_size` in the `load_state_dict` function and the definition of `indices` in the `__iter__` function to ensure that `assert len(indices) == self.num_samples`.

### Why are the changes needed?

 In the previous ElasticDistributedSampler code, the issue where `completed_num` not being divisible by `num_replicas` could cause an `AssertionError` in the scenario of importing a checkpoint was not considered.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

UT and training test.

BalaBalaYi reviewed (changes requested):
please add ut to cover the issue mentioned

sunjq1 commented:
> please add ut to cover the issue mentioned

 
since I adjusted the logic for splitting indices after loading the checkpoint, I corrected the num values in the UT file

BalaBalaYi reviewed (approved):
LGTM

zhengchenyu commented on dlrover/trainer/tests/torch/elastic_sampler_test.py:
```diff
@@ -58,7 +58,7 @@ def test_checkpoint(self):
         sampler.load_state_dict(sampler_state)
         val = next(iter(sampler))
-        self.assertEqual(val, 64)
+        self.assertEqual(val, 66)
```
This should be caused by scaling. This unit-test tells us that data will be lost.
